### ISSUE TITLE
Using quick-lint-js as library

### ISSUE THREAD

**Issue by coderaiser** [quick-lint/quick-lint-js#594]
Hi, thank you for such a great project!

I'm working on 🐊[`Putout`](https://github.com/coderaiser/putout) code transformer, and as I see, `quick-lint-js` is compiled to `wasm`.

Would be amazing to have ability to use it's API to check what is wrong with a file that cannot be parsed :).

**Comment by strager** [comment]
Can you sketch an example JavaScript API which would be ideal for Putout?

**Comment by coderaiser** [comment]
Yes! I use such concept as [`second chance`](https://github.com/coderaiser/putout/blob/v22.9.0/packages/engine-parser/lib/second-chance.js) [while parsing](https://github.com/coderaiser/putout/blob/v22.9.0/packages/engine-parser/lib/custom-parser.js#L20-L24) for things like [type assertions](https://basarat.gitbook.io/typescript/type-system/type-assertion#as-foo-vs.-less-than-foo-greater-than).

For babel is very hard to parse code like this:

```ts
var foo: any;
var bar = <string> foo; // bar is now of type "string"
```

In `typescript` + `jsx` mode.

Also when `await` used without `async`, or in case of [duplicate declaration](https://github.com/babel/babel/issues/12950) parser just crashes with no useful information.

So if you don't use concept of [`ESTree`-based](https://github.com/estree/estree) `AST`, some kind of `quickLint` (or any other name like `validate`) will be great:

```js
const quickLint = require('quick-lint-js');
const places = quickLint('() => await');
// returns
[{
    message: `'await' is only allowed in async functions`,
    position: [{
        line: 1,
        column: 1,
    }]
}];
```

This is the best possible way, but even just `message` will be much better, then `Cannot read property 'buildError' of undefined` 😄.

Also would be great to install it using `npm`.

Thank you!

**Comment by strager** [comment]
When reporting errors, how do you want quick-lint-js to count columns and lines? Is there a specific tool we should emulate?

Context: Different tools have different preferences. For example, VS Code internally counts ☃ (U+2603) as one column but 🎸 (U+1F3B8) as two columns. Vim internally counts ☃ as three columns. VS Code columns start at 0; Vim columns start at 1. JavaScript treats U+2028 and U+2029 as newlines, but VS Code and Vim don't.

**Comment by coderaiser** [comment]
The perfect way according to [ESTree](https://github.com/estree/estree/blob/master/es5.md#node-objects):

![image](https://user-images.githubusercontent.com/1573141/147074908-5926c25a-e79a-4a4c-8cf0-a8d17113f28f.png)

`@babel/parser` from https://astexplorer.net on such code:

```js
const a = '☃'; const b = 1;
```

Shows location of `VariableDeclaration` `b` as:
```js
{
    line: 1,
    column: 15
}
```

Same for such code:

```js
const a = 'x'; const b = 1;
```

So one char is one column, `columns` starts from `0`, lines starts from `1`.

**Comment by strager** [comment]
Status update:

src/quick-lint-js/c-api.h is the closest thing to a library interface. We use it for the web port: https://quick-lint-js.com/demo/ It is not exposed publicly right now (i.e. we don't ship a .wasm file with this interface); the API is unstable.

If a consumer wants more customization options today, then the `parse_and_lint` function in src/quick-lint-js/fe/linter.h might work. This is what all editor plugins use (including c-api.h). This even less stable than c-api.h.
